Code/Tools/Libs/ToolsFoundation/FileSystem/Implementation/FileSystemModel.cpp

```diff
@@ -0,0 +1,855 @@
+  {
+    ezFileStatus fileStatus;
+    ezFileStatus::Status folderStatus = ezFileStatus::Status::Unknown;
+    bool bFileExisted = false;
+    bool bFolderExisted = false;
+    {
+      EZ_LOCK(m_FilesMutex);
+      if (auto it = m_ReferencedFiles.Find(sAbsolutePath); it.IsValid())
+      {
+        bFileExisted = true;
+        fileStatus = it.Value();
+        m_ReferencedFiles.Remove(it);
+      }
+      if (auto it = m_ReferencedFolders.Find(sAbsolutePath); it.IsValid())
+      {
+        bFolderExisted = true;
+        folderStatus = it.Value();
+        m_ReferencedFolders.Remove(it);
+      }
+    }
+
+    if (bFileExisted)
+    {
+      FireFileChangedEvent(sAbsolutePath, fileStatus, ezFileChangedEvent::Type::FileRemoved);
+    }
+
+    if (bFolderExisted)
+    {
+      if (bRecurseIntoFolders)
+      {
+        ezSet<ezString> previouslyKnownFiles;
+        {
+          FILESYSTEM_PROFILE("FindReferencedFiles");
+          EZ_LOCK(m_FilesMutex);
+          auto itlowerBound = m_ReferencedFiles.LowerBound(sAbsolutePath);
+          while (itlowerBound.IsValid() && itlowerBound.Key().StartsWith(sAbsolutePath))
+          {
+            previouslyKnownFiles.Insert(itlowerBound.Key());
+            ++itlowerBound;
+          }
+        }
+        {
+          FILESYSTEM_PROFILE("HandleRemovedFiles");
+          for (const ezString& sFile : previouslyKnownFiles)
+          {
+            HandleSingleFile(sFile, false);
+          }
+        }
+      }
+      FireFolderChangedEvent(sAbsolutePath, ezFolderChangedEvent::Type::FolderRemoved);
+    }
+
+    return {};
+  }
+
+  return HandleSingleFile(sAbsolutePath, Stats, bRecurseIntoFolders);
+}
+
+
+ezFileStatus ezFileSystemModel::HandleSingleFile(const ezString& sAbsolutePath, const ezFileStats& FileStat, bool bRecurseIntoFolders)
+{
+  FILESYSTEM_PROFILE("HandleSingleFile2");
+
+  if (FileStat.m_bIsDirectory)
+  {
+    ezFileStatus status;
+    bool bExisted = false;
+    {
+      EZ_LOCK(m_FilesMutex);
+      auto it = m_ReferencedFolders.FindOrAdd(sAbsolutePath, &bExisted);
+      it.Value() = ezFileStatus::Status::Valid;
+    }
+
+    if (!bExisted)
+    {
+      FireFolderChangedEvent(sAbsolutePath, ezFolderChangedEvent::Type::FolderAdded);
+      if (bRecurseIntoFolders)
+        CheckFolder(sAbsolutePath);
+    }
+
+    return {};
+  }
+  else
+  {
+    ezFileStatus status;
+    bool bExisted = false;
+    bool bFileChanged = false;
+    {
+      EZ_LOCK(m_FilesMutex);
+      auto it = m_ReferencedFiles.FindOrAdd(sAbsolutePath, &bExisted);
+      ezFileStatus& value = it.Value();
+      bFileChanged = !value.m_Timestamp.Compare(FileStat.m_LastModificationTime, ezTimestamp::CompareMode::Identical);
+      if (bFileChanged)
+      {
+        value.m_uiHash = 0;
+      }
+
+      // If the state is unknown, we loaded it from the cache and need to fire FileChanged to update dependent systems.
+      // #TODO_ASSET This behaviors should be changed once the asset cache is stored less lossy.
```

Comment:
still todo ?

Reply:
Yep, will need to change the asset cache first to serialize the full state to cache before this quirk can be changed.